// json-schema-to-typescript/out/src/index.js
"use strict";
var pretty_printer_1 = require('./pretty-printer');
var TsTypes_1 = require('./TsTypes');
var fs_1 = require('fs');
var lodash_1 = require('lodash');
var path_1 = require('path');
var RuleType;
(function (RuleType) {
    RuleType[RuleType['Any'] = 0] = 'Any';
    RuleType[RuleType['TypedArray'] = 1] = 'TypedArray';
    RuleType[RuleType['Enum'] = 2] = 'Enum';
    RuleType[RuleType['AllOf'] = 3] = 'AllOf';
    RuleType[RuleType['AnyOf'] = 4] = 'AnyOf';
    RuleType[RuleType['Reference'] = 5] = 'Reference';
    RuleType[RuleType['NamedSchema'] = 6] = 'NamedSchema';
    RuleType[RuleType['AnonymousSchema'] = 7] = 'AnonymousSchema';
    RuleType[RuleType['String'] = 8] = 'String';
    RuleType[RuleType['Number'] = 9] = 'Number';
    RuleType[RuleType['Void'] = 10] = 'Void';
    RuleType[RuleType['Object'] = 11] = 'Object';
    RuleType[RuleType['Array'] = 12] = 'Array';
    RuleType[RuleType['Boolean'] = 13] = 'Boolean';
    RuleType[RuleType['Literal'] = 14] = 'Literal';
    RuleType[RuleType['Union'] = 15] = 'Union';
})(RuleType || (RuleType = {}));
var EnumType;
(function (EnumType) {
    EnumType[EnumType["String"] = 0] = "String";
    EnumType[EnumType["Integer"] = 1] = "Integer";
})(EnumType || (EnumType = {}));
var Compiler = (function () {
    function Compiler(schema, filePath, settings) {
        if (filePath === void 0) { filePath = ''; }
        this.schema = schema;
        var path = path_1.resolve(filePath);
        this.filePath = path_1.parse(path);
        this.declarations = new Map;
        this.id = schema.id || schema.title || this.filePath.name || 'Interface1';
        this.settings = Object.assign({}, Compiler.DEFAULT_SETTINGS, settings);
        this.declareType(this.toTsType(this.schema, '', true), this.id, this.id);
    }
    Compiler.prototype.toString = function () {
        var _this = this;
        return pretty_printer_1.format(Array.from(this.declarations.values())
            .map(function (_) { return _.toDeclaration(_this.settings); })
            .join('\n'));
    };
    Compiler.prototype.isRequired = function (propertyName, schema) {
        return schema.required ? schema.required.indexOf(propertyName) > -1 : false;
    };
    Compiler.prototype.supportsAdditionalProperties = function (schema) {
        return schema.additionalProperties === true || lodash_1.isPlainObject(schema.additionalProperties);
    };
    Compiler.prototype.getRuleType = function (rule) {
        if (rule.type === 'array' && rule.items) {
            return RuleType.TypedArray;
        }
        // enum type vs enum constant?
        if (rule.enum) {
            return RuleType.Enum;
        }
        if (rule.properties || rule.additionalProperties) {
            return RuleType.NamedSchema;
        }
        if (rule.allOf) {
            return RuleType.AllOf;
        }
        if (rule.anyOf) {
            return RuleType.AnyOf;
        }
        if (rule.$ref) {
            return RuleType.Reference;
        }
        switch (rule.type) {
            case 'array': return RuleType.Array;
            case 'boolean': return RuleType.Boolean;
            case 'integer':
            case 'number': return RuleType.Number;
            case 'null': return RuleType.Void;
            case 'object': return RuleType.Object;
            case 'string': return RuleType.String;
        }
        if (Array.isArray(rule.type)) {
            return RuleType.Union;
        }
        if (this.isNumberLiteral(rule)) {
            return RuleType.Number;
        }
        if (!lodash_1.isPlainObject(rule)) {
            return RuleType.Literal;
        }
        if (lodash_1.isPlainObject(rule)) {
            return RuleType.AnonymousSchema; // TODO: is it safe to do this as a catchall?
        }
        return RuleType.Any;
    };
    Compiler.prototype.isNumberLiteral = function (a) {
        return /^[\d\.]+$/.test(a);
    };
    // eg. "#/definitions/diskDevice" => ["definitions", "diskDevice"]
    // only called in case of a $ref type
    Compiler.prototype.resolveType = function (refPath, propName) {
        if (refPath === '#' || refPath === '#/') {
            return TsTypes_1.TsType.Interface.reference(this.id);
        }
        if (refPath[0] !== '#') {
            var id = void 0;
            var fullPath = path_1.resolve(path_1.join(this.filePath.dir, refPath));
            var file = void 0;
            if (fullPath.startsWith('http')) {
                throw new ReferenceError('Remote http references are not yet supported.  Could not read ' + fullPath);
            }
            try {
                file = fs_1.readFileSync(fullPath);
            }
            catch (err) {
                throw new ReferenceError('Unable to find referenced file ' + fullPath);
            }
            var targetType = this.toTsType(JSON.parse(file.toString()), propName, false, true);
            if (targetType.id) {
                id = targetType.toSafeType(this.settings);
            }
            else {
                var parsedNewFile = path_1.parse(fullPath);
                id = parsedNewFile.name;
            }
            if (this.settings.declareReferenced) {
                this.declareType(targetType, id, id);
            }
            return new TsTypes_1.TsType.Literal(id);
        }
        ;
        var parts = refPath.slice(2).split('/');
        var ret = this.settings.declareReferenced ? this.declarations.get(parts.join('/')) : undefined;
        if (!ret) {
            var cur = this.schema;
            for (var i = 0; cur && i < parts.length; ++i) {
                cur = cur[parts[i]];
            }
            ret = this.toTsType(cur);
            if (this.settings.declareReferenced && (this.settings.declareSimpleType || !ret.isSimpleType()))
                this.declareType(ret, parts.join('/'), this.settings.useFullReferencePathAsName ? parts.join('/') : lodash_1.last(parts));
        }
        return ret;
    };
    Compiler.prototype.declareType = function (type, refPath, id) {
        type.id = id;
        this.declarations.set(refPath, type);
        return type;
    };
    Compiler.prototype.toStringLiteral = function (a) {
        var _this = this;
        switch (typeof a) {
            case 'boolean': return new TsTypes_1.TsType.Boolean; // ts doesn't support literal boolean types
            case 'number': return new TsTypes_1.TsType.Number; // ts doesn't support literal numeric types
            case 'string': return new TsTypes_1.TsType.Literal(JSON.stringify(a));
            default: return new TsTypes_1.TsType.Interface(lodash_1.map(a, function (v, k) {
                return {
                    name: k,
                    required: true,
                    type: _this.toStringLiteral(v)
                };
            }));
        }
    };
    Compiler.prototype.createTsType = function (rule, propName, isTop, isReference) {
        var _this = this;
        if (isTop === void 0) { isTop = false; }
        if (isReference === void 0) { isReference = false; }
        switch (this.getRuleType(rule)) {
            case RuleType.AnonymousSchema:
            case RuleType.NamedSchema:
                return this.toTsDeclaration(rule);
            case RuleType.Enum:
                // we honor the schema's "type" on the enum.  if string, generate a union.
                // if int, require the tsEnumNames
                var enumType = this.validateEnumMembers(rule);
                switch (enumType) {
                    case EnumType.Integer:
                        var enumValues = lodash_1.zip(rule.tsEnumNames || [], 
                        // If we try to create a literal from an object, bad stuff can happen... so we have to toString it
                        rule.enum.map(function (_) { return new TsTypes_1.TsType.Literal(_).toType(_this.settings).toString(); }))
                            .map(function (_) { return new TsTypes_1.TsType.EnumValue(_); });
                        // name our anonymous enum, if it doesn't have an ID, by the property name under
                        // which it was declared.  Failing both of these things, it'll concat together the
                        // identifiers as EnumOneTwoThree for enum: ["One", "Two", "Three"].  Ugly, but
                        // practical.
                        var path = rule.id || propName || ('Enum' + enumValues.map(function (_) { return _.identifier; }).join(''));
                        var enm = new TsTypes_1.TsType.Enum(enumValues);
                        var retVal = enm;
                        // don't add this to the declarations map if this is the top-level type (already declared)
                        // or if it's a reference and we don't want to declare those.
                        if ((!isReference || this.settings.declareReferenced)) {
                            if (!isTop) {
                                retVal = this.declareType(retVal, path, path);
                            }
                            else {
                                retVal.id = path;
                            }
                        }
                        return retVal;
                    case EnumType.String:
                        return new TsTypes_1.TsType.Union(lodash_1.uniqBy(rule.enum.map(function (_) { return _this.toStringLiteral(_); }), function (_) { return _.toType(_this.settings); }));
                }
            case RuleType.Any: return new TsTypes_1.TsType.Any;
            case RuleType.Literal: return new TsTypes_1.TsType.Literal(rule);
            case RuleType.TypedArray: return new TsTypes_1.TsType.Array(this.toTsType(rule.items));
            case RuleType.Array: return new TsTypes_1.TsType.Array;
            case RuleType.Boolean: return new TsTypes_1.TsType.Boolean;
            case RuleType.Number: return new TsTypes_1.TsType.Number;
            case RuleType.Object: return new TsTypes_1.TsType.Object;
            case RuleType.String: return new TsTypes_1.TsType.String;
            case RuleType.Void: return new TsTypes_1.TsType.Void;
            case RuleType.AllOf:
                return new TsTypes_1.TsType.Intersection(rule.allOf.map(function (_) { return _this.toTsType(_); }));
            case RuleType.AnyOf:
                return new TsTypes_1.TsType.Union(rule.anyOf.map(function (_) { return _this.toTsType(_); }));
            case RuleType.Reference:
                return this.resolveType(rule.$ref, propName);
            case RuleType.Union:
                return new TsTypes_1.TsType.Union(rule.type.map(function (_) { return _this.toTsType(_); }));
        }
        throw new Error('Unknown rule:' + rule.toString());
    };
    Compiler.prototype.validateEnumMembers = function (rule) {
        if (!rule.type)
            rule.type = 'string';
        var isDeclaredStringEnum = rule.type === 'string';
        var isDeclaredIntegerEnum = rule.type === 'integer';
        if (!isDeclaredStringEnum && !isDeclaredIntegerEnum) {
            throw TypeError('Enum type must be string or integer; default is string if undefined');
        }
        if (rule.enum.some(function (_) { return _ instanceof Object; })) {
            throw TypeError('Enum members must be a list of strings or a list of integers; instead, found an Object');
        }
        var isActuallyStringEnum = rule.enum.every(function (_) { return typeof (_) === 'string'; });
        var isActuallyIntegerEnum = rule.enum.every(function (_) { return typeof (_) === 'number'; });
        var isIntegerEnumWithValidStringValues = isActuallyIntegerEnum
            && rule.tsEnumNames
            && rule.tsEnumNames.length === rule.enum.length
            && rule.tsEnumNames.every(function (_) { return typeof (_) === 'string'; });
        if (isDeclaredStringEnum && !isActuallyStringEnum) {
            throw TypeError('Enum was declared as a string type but found at least one non-string member');
        }
        if (isDeclaredIntegerEnum && !isIntegerEnumWithValidStringValues) {
            if (!isActuallyIntegerEnum) {
                throw TypeError('Enum was declared as an integer type, but found at least one non-integer member');
            }
            if (!rule.tsEnumNames) {
                throw TypeError('Property tsEnumNames is required when enum is declared as an integer type');
            }
            if (rule.tsEnumNames.length !== rule.enum.length) {
                throw TypeError('Property enum and property tsEnumNames must be the same length');
            }
            throw TypeError('Enum was declared as an integer type, but found at least one non-string tsEnumValue');
        }
        // I don't think we should ever hit this case.
        if (!isActuallyStringEnum && !isIntegerEnumWithValidStringValues) {
            throw TypeError('Enum members must be a list of strings or a list of integers (with corresponding tsEnumNames)');
        }
        if (isIntegerEnumWithValidStringValues) {
            return EnumType.Integer;
        }
        else {
            return EnumType.String;
        }
    };
    Compiler.prototype.toTsType = function (rule, propName, isTop, isReference) {
        if (isTop === void 0) { isTop = false; }
        if (isReference === void 0) { isReference = false; }
        var type = this.createTsType(rule, propName, isTop, isReference);
        if (!type.id) {
            // the type is not declared, let's check if we should declare it or keep it inline
            type.id = rule.id || rule.title;
            if (type.id && !isReference)
                this.declareType(type, type.id, type.id);
        }
        type.description = type.description || rule.description;
        return type;
    };
    Compiler.prototype.toTsDeclaration = function (schema) {
        var _this = this;
        var copy = lodash_1.merge({}, Compiler.DEFAULT_SCHEMA, schema);
        var props = lodash_1.map(copy.properties, function (v, k) {
            return {
                name: k,
                required: _this.isRequired(k, copy),
                type: _this.toTsType(v, k)
            };
        });
        if (props.length === 0 && !('additionalProperties' in schema)) {
            if ('default' in schema)
                return new TsTypes_1.TsType.Void;
        }
        if (this.supportsAdditionalProperties(copy)) {
            var short = copy.additionalProperties === true;
            if (short && props.length === 0)
                return new TsTypes_1.TsType.Any;
            var type = short ? new TsTypes_1.TsType.Any : this.toTsType(copy.additionalProperties);
            props.push({
                name: '[k: string]',
                required: true,
                type: type
            });
        }
        return new TsTypes_1.TsType.Interface(props);
    };
    Compiler.DEFAULT_SETTINGS = TsTypes_1.TsType.DEFAULT_SETTINGS;
    Compiler.DEFAULT_SCHEMA = {
        additionalProperties: true,
        properties: {},
        required: [],
        type: 'object'
    };
    return Compiler;
}());
function compile(schema, path, settings) {
    return new Compiler(schema, path, settings).toString();
}
exports.compile = compile;
function compileFromFile(inputFilename) {
    return new Promise(function (resolve, reject) {
        return fs_1.readFile(inputFilename, function (err, data) {
            if (err) {
                reject(err);
            }
            else {
                resolve(compile(JSON.parse(data.toString()), inputFilename));
            }
        });
    });
}
exports.compileFromFile = compileFromFile;
