// json-schema-to-typescript/out/test/cases/basics.js
"use strict";
exports.schema = {
    "title": "Example Schema",
    "type": "object",
    "properties": {
        "firstName": {
            "type": "string"
        },
        "lastName": {
            "type": "string"
        },
        "age": {
            "description": "Age in years",
            "type": "integer",
            "minimum": 0
        },
        "height": {
            "type": "number"
        },
        "favoriteFoods": {
            "type": "array"
        },
        "likesDogs": {
            "type": "boolean"
        }
    },
    "required": ["firstName", "lastName"]
};
exports.configurations = [
    {
        settings: {
            useInterfaceDeclaration: true,
        },
        types: "export interface ExampleSchema {\n  firstName: string;\n  lastName: string;\n  /** Age in years */\n  age?: number;\n  height?: number;\n  favoriteFoods?: any[];\n  likesDogs?: boolean;\n  [k: string]: any;\n}"
    },
    {
        settings: {
            propertyDescription: false
        },
        types: "export interface ExampleSchema {\n  firstName: string;\n  lastName: string;\n  age?: number;\n  height?: number;\n  favoriteFoods?: any[];\n  likesDogs?: boolean;\n  [k: string]: any;\n}"
    }
];
